mcs/class/corlib/System.Reflection.Emit/ParameterBuilder.cs

```diff
@@ -96,12 +96,24 @@ public virtual void SetConstant (object defaultValue)
 		{
 			if (position > 0) {
 				Type t = methodb.GetParameterType (position - 1);
-				if (defaultValue != null && t != defaultValue.GetType ()) {
-					if(!t.IsEnum || t.UnderlyingSystemType != defaultValue.GetType ())
-						throw new ArgumentException ("Constant does not match the defined type.");
+				if (defaultValue == null) {
+					if (!(t.IsGenericType && t.GetGenericTypeDefinition () == typeof(Nullable<>)))
+						throw new ArgumentException("Null is not a valid constant value for this type");
+				} else {
+					if (t.IsEnum) {
+						Type underlyingType = Enum.GetUnderlyingType (t);
```

Comment:
This check might fail if `t` is `Nullable<SomeEnum>`. You would then be unable to set the default value for a nullable enum parameter.

You might have to "unwrap" nullable types before the check, e.g.:

```csharp
var unpackedIfNullableT = Nullable.GetUnderlyingType(t) ?? t;
if (unpackedIfNullableT.IsEnum) ...
```